### DISCUSSION TITLE
Nvidia not working after a few days

### DISCUSSION THREAD

**Discussion by Cah1982** [blakeblackshear/frigate#7915] [General]
So my setup works but after a few days I see this

![image](https://github.com/blakeblackshear/frigate/assets/44786907/04d15933-0db7-4a6b-9b18-1c579f07b992)

In frigate ssh i get
![image](https://github.com/blakeblackshear/frigate/assets/44786907/2aac6cd3-94f5-4a61-bd46-6418be786161)

My Jellyfin also uses the nvidia drivers but shows
![image](https://github.com/blakeblackshear/frigate/assets/44786907/46641b25-c302-4adc-a394-173fcf0b1920)

If i simply restart the Frigate container it works again
![image](https://github.com/blakeblackshear/frigate/assets/44786907/f60e0fb4-4067-490d-8bac-922fc7d253bb)

![image](https://github.com/blakeblackshear/frigate/assets/44786907/c3bd497c-2d3b-4058-af31-46b3bafb72c1)

So how would I work out why its doing it to make this stable. Jellyfin never needs restarting

**Comment by vajonam** [comment]
I too see this from time to time, at the moment my sledgehammer method is to restart everyday. I have turned off the auto restart with the beta, to see if can catch it and provide some logs to help debug.

**Reply by Cah1982** [reply]
yeah, seems random, could be a day, could be 3-4 days. but jellyfin never needs a restart, only time thats needed is after a system update,

**Comment by vajonam** [comment]
https://github.com/NVIDIA/nvidia-docker/issues/1671

**Comment by vajonam** [comment]
This seems like a reasonable work around, not a sledge hammer, but a slightly less intrusive hammer!  This should affect all nvidia based containers. not sure why it doesn't affect jellyfin, i do see it effect plex, but I dont use jellyfin as much so I cannot confirm it doesnt affect jf. 

https://stackoverflow.com/a/73704403

**Comment by vajonam** [comment]
The fix works nicely, just tried it.. i.e the health-check and marking the container dirty. I didn't want to futz with all the cgroup work around seems a lot more intrusive. 

@NickM-27 @blakeblackshear  is it possible to add this health check into the either to tensorrt image, or to the documentation on launching the tensorrt image. I am sure this effects anyone using NVIDIA gpu's through a container on a debian or similar system that also uses systemd for init. Its so intermittent its hard to diagnose and address. 

easy to reproduce if do `systemctl daemon-reload` on the host you will notice all the conatiners that need GPU access stop working (including jellyfin,plex etc.).

**Reply by NickM-27** [reply]
> I am sure this effects anyone using NVIDIA gpu's through a container. Its so intermittent its hard to diagnose and address.

I have had no such issues and I have used an nvidia GPU for over a year now

**Reply by NickM-27** [reply]
this is also the first time I have seen this issue reported, to be clear

**Comment by vajonam** [comment]
you can test it out by doing `systemctl daemon-reload` you might have just gotten lucky if you dont really do other work on that system and just chugs along.  if you do a apt update or anything that causes a reload it should break.

**Reply by vajonam** [reply]
its been happening to me for a long time now, I just used to restart daily to not run into it.

**Reply by NickM-27** [reply]
it is my NAS so I am constantly doing things, but it is unRAID which is based on slackware linux. I have many different things running on the nvidia GPU and never had any issues. 

In any case, if we are going to make a recommendation I think it is better to understand why and when that recommendation applies instead of recommending every nvidia GPU user to do it when it only applies to a small subset. Given the lack of issue reports for this, it implies to me this is not even as popular as every user running debian / ubuntu based OS for example.

**Comment by vajonam** [comment]
No worries can just watch it.. I am so happy I found an fix for it and it happened to someone else! Never nice when youre the only having the issue.   Suspect these issues are related. 

https://github.com/blakeblackshear/frigate/issues/5650

**Reply by NickM-27** [reply]
So every camera goes offline? it isn't clear to me what the symptoms are either, the only thing mentioned so far is nvidia-smi not working

**Reply by vajonam** [reply]
yeah any camera using Nvidia GPU hw decoding stops works (in my case all cameras, 5 of them), and object detection process stops working. but because the cameras aren't sending any frame, there arent any detection errors in the log, just ones related related to decoding and ffmpeg.

**Reply by vajonam** [reply]
`nvidia-smi` not working is the easiest symptom to inspect in the container. that's why that is used.

**Comment by vajonam** [comment]
Just to close the loop on this, I have also tried the steps outlined here.

https://github.com/NVIDIA/nvidia-docker/issues/1730

Explicitly disabling systemd cgroup management in Docker

- Set the parameter `"exec-opts": ["native.cgroupdriver=cgroupfs"]` in the `/etc/docker/daemon.json` file and restart docker.

This seems to fix the issue where a `sytemed daemon-reload` causes the containers to loose access to the GPU. 

To catch any other scenario which might cause this and are fixed by a restart, I am also using a health check as part of my docker compose for the frigate container. 

```
      healthcheck:
        test: ["CMD-SHELL", "test -s `which nvidia-smi` && nvidia-smi || exit 1"]
        start_period: 1s
        interval: 20s
        timeout: 5s
        retries: 2
      labels:
        - autoheal=true
        - autoheal.stop.timeout=1

```

and also add container beside frigate 

```
    autoheal:
      image: willfarrell/autoheal
      container_name: autoheal
      environment:
        - AUTOHEAL_CONTAINER_LABEL=all
      volumes:
        - /var/run/docker.sock:/var/run/docker.sock
      restart: always

``` 
which will restart the container on a unhealthy result from the check.  Just for good measure I added a depends on in the frigate container. 
```
      depends_on:
        - autoheal
```
that ensure autoheal is running before frigate is started.
